Syndication/AzureStack.MarketplaceSyndication.psm1

```diff
@@ -90,165 +83,1086 @@ function Sync-AzSOfflineMarketplaceItem {
+            Get-Dependency -productid $id -resourceGroup $resourceGroup
+        }
+    }
 
-            If ($FileExists -eq $true) {Remove-Item "$destination\$azpkgName.txt" -force} else {
-                New-Item "$destination\$azpkgName.txt"
-            }
-            $productDetails.properties|select publisherIdentifier,offer,sku,productKind,vmExtensionType  |out-file "$destination\$azpkgName.txt" -Append
-            $productDetails.properties.productProperties|select version| out-file "$destination\$azpkgName.txt" -Append
+    Write-Host "`nDownloading product: $productid" -ForegroundColor DarkCyan
+    Download-Product -productid $productid -resourceGroup $resourceGroup
+}
 
+function Download-Product {
+    param (
+        [parameter(mandatory = $true)]
+        [String] $productid,
+
+        [parameter(mandatory = $true)]
+        [String] $resourceGroup
+    )
+
+    # get name of azpkg
+    $azpkgURI = "$($azureEnvironment.ResourceManagerUrl.ToString().TrimEnd('/'))/subscriptions/$($AzureSubscriptionID.ToString())/resourceGroups/$resourceGroup/providers/Microsoft.AzureStack/registrations/$Registration/products/$($productid)?api-version=2016-01-01"
+    Write-Debug $azpkgURI
+    $Headers = @{ 'authorization' = "Bearer $($Token.AccessToken)"}
+    $productDetails = Invoke-RestMethod -Method GET -Uri $azpkgURI -Headers $Headers
+    $azpkgName = $productDetails.properties.galleryItemIdentity
+    if (!$azpkgName) {
+        $azpkgName = $productid
+    }
+
+    # get download location for azpkg
+    $downloadURI = "$($azureEnvironment.ResourceManagerUrl.ToString().TrimEnd('/'))/subscriptions/$($AzureSubscriptionID.ToString())/resourceGroups/$resourceGroup/providers/Microsoft.AzureStack/registrations/$Registration/products/$productid/listDetails?api-version=2016-01-01"
+    Write-Debug $downloadURI
+    $downloadDetails = Invoke-RestMethod -Method POST -Uri $downloadURI -Headers $Headers
+
+    # create Legal Terms POPUP
+    $legalTitle = "Legal Terms"
+    $legalText = $productDetails.properties.description -replace '<[^>]+>',''
+    Write-Host $("-"*20)
+    Write-Host "$legalTitle`n$legalText" -ForegroundColor DarkYellow
+    $confirmation = Read-Host "Accept Legal Terms. (Y/N)?"
+    If ($confirmation -eq 'Y') {
+        $productFolder = "$destination\$productid"
+
+        # output parameters required for import
+        $FileExists = Test-Path "$productFolder\$azpkgName.txt"
+        $DestinationCheck = Test-Path $productFolder
+        If ($DestinationCheck -eq $false) {
+            New-item -ItemType Directory -force $productFolder | Out-Null
+        }
+
+        If ($FileExists -eq $true) {Remove-Item "$productFolder\$azpkgName.txt" -force -ErrorAction SilentlyContinue | Out-Null} else {
+            New-Item "$productFolder\$azpkgName.txt" | Out-Null
+        }
+
+        If ($FileExists -eq $true) {Remove-Item "$productFolder\$azpkgName.json" -force -ErrorAction SilentlyContinue | Out-Null}
+
+        $productInfo = @{}
```

Comment:
this code can be "$productDetails.properties | Select displayName, description, etc." but why are you creating a new object at all here?

Reply:
Want to use $productInfo to capture the metadata of product, then dump to a json file, which can be used later in import stage. AzureBridge API needs the saved metadata.